storage/dataflux/fast_list.go

```diff
@@ -127,35 +132,58 @@ func NewLister(c *storage.Client, in *ListerInput) *Lister {
 // worksteal listing is expected to be faster.
 func (c *Lister) NextBatch(ctx context.Context) ([]*storage.ObjectAttrs, error) {
 	// countError tracks the number of failed listing methods.
-	countError := 0
+	cc := &contextErr{counter: 0}
+
 	var results []*storage.ObjectAttrs
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
+
+	wsCompletedfirst := false
+	seqCompletedfirst := false
+	var wsObjects []*storage.ObjectAttrs
+	var seqObjects []*storage.ObjectAttrs
+	var nextToken string
+
 	// Errgroup takes care of running both methods in parallel. As soon as one of
 	// the method is complete, the running method also stops.
 	g, childCtx := errgroup.WithContext(ctx)
 
 	// To start listing method is Open and runs both worksteal and sequential listing
 	// in parallel. The method which completes first is used for all subsequent runs.
 
-	// TODO: Run worksteal listing when method is Open or WorkSteal.
+	// Run worksteal listing when method is Open or WorkSteal.
+	if c.method != sequential {
```

Comment:
I think this method would be easier to understand with a little repetition, using a switch on `c.method`. That way we also don't have to create child contexts + errgroups every time this runs, and the common/simple case (where we already know what kind of algorithm we're running) is easy to parse quickly.

Here's an outline of what I am thinking:

```go
	switch c.method {
	case sequential:
		objects, token, err := c.sequentialListing(ctx)
		if err != nil {
			return nil, fmt.Errorf("error in running sequential listing: %w", err)
		}
		c.pageToken = token
		
		// do checks and return here, or save objects in the results var to do the
		// checks at the end
		if c.pageToken == "" {
		   return objects, iterator.Done
	        }
		
		return objects, nil // or save results = objects
	case worksteal:
		objects, err := c.workstealListing(ctx)
		if err != nil {
			return nil, fmt.Errorf("error in running worksteal_lister: %w", err)
		}
		return objects, nil // add check too, or same as above, save results
	case open:
                // Do both here:
                // We can move all vars that are only relevant to this case here
                	ctx, cancel := context.WithCancel(ctx)
		defer cancel()
		wsCompletedfirst := false
		seqCompletedfirst := false
		var wsObjects []*storage.ObjectAttrs
		var seqObjects []*storage.ObjectAttrs
		var nextToken string
	
		g, childCtx := errgroup.WithContext(ctx)
		g.Go(func() error {
			objects, err := c.workstealListing(childCtx)
			// and so on...
		})

		g.Go(func() error {
			objects, token, err := c.sequentialListing(childCtx)
			// etc
		})

		// Close all functions if either sequential listing or worksteal listing is complete.
		err := g.Wait()
		
		// Now we only need to do these checks when the method type is open
		if err != nil && (!strings.Contains(err.Error(), context.Canceled.Error()) || cc.counter > 1) {
			return nil, fmt.Errorf("failed waiting for sequential and work steal lister : %w", err)
		}
		if wsCompletedfirst {
			//...
		} else if seqCompletedfirst {
			//...
		}
	}
```

Reply:
Done.